Autosaves per Game Name - Quality of Life Enhancement

mattthewaz opened the pull request:
When I play multiplayer games, I often cycle between files frequently depending on who I'm playing with. I have a file here that I play with one friend, another there that I player with a couple others, so on and so forth. With all this switching around, it can be easy to forget to save a Replay and have all traces of that file overwritten as your last hope - the autosaves - are overwritten by the other games you're hosting.
This change solves that by including the name of the game session in the autosave's name. Each name will support the usual 5 slots as per normal. For example, if I host a game and give it the name "Together", then the autosaves created from this will be named "Autosave-Together-1", "Autosave-Together-2", etc... and will not overwrite other autosaves from sessions with other names.
Note: the autosave's use the name of the game session (the name picked in the host dialog), not the name of the loaded file/replay.

![autosavePerGameName](https://user-images.githubusercontent.com/737914/55600706-38d1e200-572b-11e9-8724-abadd7fe7668.png)

mattthewaz committed:
Autosaves per Game Name - Quality of Life Enhancement
When I play multiplayer games, I often cycle between files frequently depending on who I'm playing with. I have a file here that I play with one friend, another there that I player with a couple others, so on and so forth. With all this switching around, it can be easy to forget to save a Replay and have all traces of that file overwritten as your last hope - the autosaves - are overwritten by the other games you're hosting.
This change solves that by including the name of the game session in the autosave's name. Each name will support the usual 5 slots as per normal. For example, if I host a game and give it the name "Together", then the autosaves created from this will be named "Autosave-Together-1", "Autosave-Together-2", etc... and will not overwrite other autosaves from sessions with other names.
Note: the autosave's use the name of the game session (the name picked in the host dialog), not the name of the loaded file/replay.

mattthewaz commented:
For testing, I did confirm that it still obeys the 5 autosave limit; it will find the oldest autosave with a matching name to overwrite when it runs out of slots.

jacobEAdamson commented:
I'm torn because autosaves currently work like they do in vanilla. Let me talk with some of the other devs to make sure we want this feature. Thanks for coding it by the way.

mattthewaz commented:
If nothing else a toggle can be added for this behavior.

jacobEAdamson commented:
Yeah, if you can gate it behind a setting in the multiplayer mod, that would be good

tindolt-zz commented:
Add a setting to turn that on and off and I agree it would be kinda nice.

tindolt-zz reviewed (changes requested):
Add mod setting toggle for turning this feature on and off and it should be good, have it default to off.

mattthewaz committed:
Added toggle that controls whether or not the game name will be appended to the autosave names. The toggle is set by the host from the HostWindow and defaults to off to maintain the default user experience.
Added new entry to the language files for this new option. The same value is provided for each (in English) so these will need to be localized later.

mattthewaz commented:
I have added the requested toggle to the Host dialog, defaulted it to 'off' to maintain the vanilla user experience, and added an entry to each Language file for the new option.

Manually tested and confirmed that the toggle works as expected.

tindolt-zz reviewed (approved):
Looks great! I am sure several people will like this feature.

MilanDierick committed:
Merge branch 'master' into autosavesPerGameName

MilanDierick commented:
I resolved a merge conflict in the German translation file.

SamboyCoding reviewed (approved):
Looks good to me!

MilanDierick reviewed (dismissed):
Please merge master branch into autosavesPerGameName branch, I just merged tons of code refactoring.

mattthewaz committed:
Merge pull request #1 from Parexy/master

Pulling master into autosavePerGameName

mattthewaz commented:
@MilanDierick , I have merged master into the autosavesPerGameName, as requested.

MilanDierick commented:
@mattthewaz Another issue would be that most people just leave the name as default when hosting, which means when a player hosts multiple games with the same name, how do you differentiate between the different auto saves? I propose that, in case the player chose to use the default name, to keep track of the auto saves, and when the player saves the replay, to rename the auto saves so they no longer use the default name, and thus prevent auto save name clashes when playing multiple multiplayer games at the same time.

mattthewaz commented:
I agree that my solution would not help people who do not change the default name; some level of user action is involved if they want to use the feature effectively. Here are the options as I see them:

- Change nothing for now. Additional rules constitute scope creep and might over-complicate the feature. 
- Follow your suggestion. We would detect when the Host has used the default name and use the Replay name when creating the filenames. In case the game wasn't loaded from Replay, still use the default name in the autosave.
- Simplified version of your suggestion. We could always use the Replay name if one exists and only use the game name when one does not.

mattthewaz commented:
I personally feel like the enhancement as it stands is a good starting point. It would be easy to iterate on this later.

jacobEAdamson commented on Source/Client/Windows/HostWindow.cs:
```diff
@@ -104,6 +104,13 @@ public override void DoWindowContents(Rect inRect)
 
             entry = entry.Down(30);
 
+            var appendNameToAutosaveLabel = $"{"MpAppendNameToAutosave".Translate()}:  ";
```
So, mostly I like the idea, but I'd really prefer not to clutter the host page with this option. The host game screen should only be for options that make sense on a session by session basis. Is it possible to move this to the mod settings located in the `MultiplayerMod.cs` file?

mattthewaz replied:
@jacobEAdamson I agree, moving this out of the Host dialog and into Mod settings sounds like a good idea.

mattthewaz replied:
I'll work on that when I get home later tonight

mattthewaz replied:
This change has been made.
![autosavePerGameNameModSettings](https://user-images.githubusercontent.com/737914/55660013-324a7580-57d2-11e9-8ac7-d942a547b8b9.png)

mattthewaz committed:
Moved the appendNameToAutosave option to the Mod Setting page.

mattthewaz committed:
Merging in upstream/master

mattthewaz commented:
Addressed merge conflicts.

MilanDierick reviewed (approved):
Looks good to me.

mattthewaz commented:
I have targeted the pull request to go to development, as per the process
